### PR TITLE
Fix __eq__ and __getitem__

### PR DESCRIPTION
**Pull request by wetneb**
See #18 and #19.

Implementation of `__lt__` could potentially be made more efficient - in doubt, I have used the reference implementation, so at least it should be correct.

### CODE REVIEW

File: sparse_list.py [hunk 1]
```diff
@@ -55,12 +55,14 @@ def __setitem__(self, index, value):
...
         try:
             s = slice(index.start, index.stop, index.step).indices(self.size)
-            return SparseList(
+            sl = SparseList(
                 {
                     k: self.elements[i]
                     for k, i in enumerate(xrange(*s))
                     if i in self.elements
                 }, self.default)
+            sl.size = len(xrange(*s))
```
Comment (johnsyweb) [inline sparse_list.py]: It feels like we should be able to calculate `sl.size` in constant time using a function of `index.start, index.stop & index.step`. See https://stackoverflow.com/a/36188683/78845 for example.

File: sparse_list.py [hunk 2]
```diff
@@ -144,13 +146,13 @@ def __initialise_from_iterable(self, arg):
         self.size = len(self.elements)
 
     def __eq__(self, other):
-        return all(a == b for a, b in zip_longest(self, other))
+        return len(self) == len(other) and all(a == b for a, b in zip_longest(self, other))
```
Comment (johnsyweb) [inline sparse_list.py]: Nice optimisation.

File: sparse_list.py [hunk 3]
```diff
@@ -144,13 +146,13 @@ def __initialise_from_iterable(self, arg):
...
 
     def __ne__(self, other):
         return not self.__eq__(other)
 
     def __lt__(self, other):
-        return any(a < b for a, b in zip_longest(self, other))
+        return list(self) < list(other)
```
Comment (johnsyweb) [inline sparse_list.py]: It's unclear which unit test this fixes.

Could we not short-circuit the evaluation on size of self _vs_ other and keep the rest of the evaluation lazy?

Comment (wetneb) [reply sparse_list.py]: The problem is that `any(a < b for a, b in zip_longest(self, other))` is not the lexical order that `list` uses… Let me add a test for that.

File: sparse_list.py (hunk 1 above)
Comment (wetneb) [reply sparse_list.py]: Yes that's probably possible, but it's also likely that we would miss some edge cases…

Comment (wetneb) [reply sparse_list.py]: Actually I am wondering if Python does not already do the computation of``len(range(*s))` in constant time, given https://stackoverflow.com/a/36188948/985087

Comment (wetneb) [reply sparse_list.py]: Yes it seems that this is already done in constant time by Python, without enumerating all the indices. (Confirmed by testing `timeit`).

File: test_sparse_list.py [hunk 4]
```diff
@@ -264,6 +264,22 @@ def test_greater_than(self):
         self.assertFalse(b <= a)
         self.assertFalse(b < a)
 
+    def test_less_than_2(self):
+        a = sparse_list.SparseList([1,2,3])
+        b = sparse_list.SparseList([1,0,4])
+        self.assertFalse(a < b)
```
Comment (wetneb) [inline test_sparse_list.py]: This test shows why the implementation of `__lt__` with `any` is incorrect.

**Comment by wetneb** [comment]
@johnsyweb let me know if there are any outstanding issues with this PR.

**Comment by wetneb** [comment]
Hi @johnsyweb, here is another friendly ping :)

File: sparse_list.py (hunk 3 above)
Comment (johnsyweb) [reply sparse_list.py]: Thanks for the test, that made it a lot clearer.

Converting two whole SparseLists to lists for this comparison seems expensive. Should we do this instead?

```suggestion
        for a, b in zip_longest(self, other):
            if a < b:
                return True
            if a > b:
                return False
        return False
```

File: test_sparse_list.py [hunk 5]
```diff
@@ -249,6 +264,22 @@ def test_greater_than(self):
         self.assertFalse(b <= a)
```
Comment (johnsyweb) [inline test_sparse_list.py]: I should break this out into its own test for less-then-or-equal-to.

File: test_sparse_list.py [hunk 6]
```diff
@@ -233,6 +238,16 @@ def test_inequality_left_longer(self):
...
 
+    def test_inequality_length(self):
+        a = sparse_list.SparseList(2)
+        b = sparse_list.SparseList(4)
+        self.assertTrue(a != b)
+        self.assertTrue(not (a == b))
+        self.assertNotEqual(a, b)
+        self.assertTrue(b != a)
+        self.assertTrue(not (b == a))
+        self.assertNotEqual(b, a)
```
Comment (johnsyweb) [inline test_sparse_list.py]: 😍

File: test_sparse_list.py [hunk 7]
```diff
@@ -249,6 +264,22 @@ def test_greater_than(self):
         self.assertFalse(b <= a)
         self.assertFalse(b < a)
 
+    def test_less_than_2(self):
```
Comment (johnsyweb) [inline test_sparse_list.py]:
```suggestion
    def test_less_than_with_a_pair_that_is_greater(self):
```

File: test_sparse_list.py [hunk 8]
```diff
@@ -249,6 +264,22 @@ def test_greater_than(self):
...
+        a = sparse_list.SparseList([1,2,3])
```
Comment (johnsyweb) [inline test_sparse_list.py]:
```suggestion
        a = sparse_list.SparseList([1, 2, 3])
```

File: test_sparse_list.py [hunk 9]
```diff
@@ -249,6 +264,22 @@ def test_greater_than(self):
...
+        b = sparse_list.SparseList([1,0,4])
```
Comment (johnsyweb) [inline test_sparse_list.py]:
```suggestion
        b = sparse_list.SparseList([1, 0, 4])
```

File: sparse_list.py [hunk 10]
```diff
@@ -55,12 +55,14 @@ def __setitem__(self, index, value):
     def __getitem__(self, index):
         try:
             s = slice(index.start, index.stop, index.step).indices(self.size)
-            return SparseList(
+            sl = SparseList(
                 {
                     k: self.elements[i]
                     for k, i in enumerate(xrange(*s))
```
Comment (johnsyweb) [inline sparse_list.py]:
```suggestion
                    for k, i in enumerate(indices)
```

File: sparse_list.py (hunk 1 above)
Comment (johnsyweb) [reply sparse_list.py]:
```suggestion
            sl.size = len(indexes)
```

File: sparse_list.py [hunk 11]
```diff
@@ -144,13 +147,18 @@ def __initialise_from_iterable(self, arg):
         self.size = len(self.elements)
 
     def __eq__(self, other):
-        return all(a == b for a, b in zip_longest(self, other))
+        return len(self) == len(other) and all(a == b for a, b in zip_longest(self, other))
```
Comment (johnsyweb) [inline sparse_list.py]: Having asserted that the two sequences are the same length, we can remove the call to `zip_longest()`

```suggestion
        return len(self) == len(other) and all(a == b for a, b in zip(self, other))
```

Comment (johnsyweb) [reply sparse_list.py]: And this means we can remove `from six.moves import zip_longest`

File: sparse_list.py [hunk 12]
```diff
@@ -144,13 +147,18 @@ def __initialise_from_iterable(self, arg):
...
         return not self.__eq__(other)
 
     def __lt__(self, other):
-        return any(a < b for a, b in zip_longest(self, other))
+        for a, b in zip(self, other):
+            if a < b:
+                return True
+            if a > b:
+                return False
+        return len(self) < len(other)
```
Comment (johnsyweb) [inline sparse_list.py]: Aha. Good call!

**Comment by johnsyweb** [comment]
Thank you very much for this contribution, @wetneb 🙇
